dandi/dandiapi.py

```diff
@@ -729,13 +832,61 @@ def api_path(self) -> str:
     def download_url(self) -> str:
         return self.client.get_url(f"{self.api_path}download/")
 
+    def get_metadata(self) -> models.Asset:
+        """
+        Fetch the metadata for the asset as a `dandischema.models.Asset`
+        instance
+        """
+        return models.Asset.parse_obj(self.get_raw_metadata())
+
     def get_raw_metadata(self) -> Dict[str, Any]:
         """Fetch the metadata for the asset as an unprocessed `dict`"""
         if self._metadata is not None:
             return self._metadata
         else:
             return cast(Dict[str, Any], self.client.get(self.api_path))
 
+    def get_digest(self, digest_type: Union[str, models.DigestType]) -> str:
+        """
+        Retrieves the value of the given type of digest from the asset's
+        metadata.  Raises `NotFoundError` if there is no entry for the given
+        digest type.
+        """
+        if isinstance(digest_type, models.DigestType):
+            digest_type = digest_type.value
+        metadata = self.get_raw_metadata()
+        try:
+            return metadata["digest"][digest_type]
+        except KeyError:
+            raise NotFoundError(f"No {digest_type} digest found in metadata")
+
+    def set_metadata(self, metadata: models.Asset) -> None:
```

Comment:
I think such a name would be quite a bit off with our "Copy-on-Write" implementation of dandi-api: metadata cannot really be modified, but rather we provide a new metadata and get a new asset (id).  So `set_raw_metadata` seems to be doing that below and returns a new `RemoteAsset`.  I wonder if we should just promote it into `get_modified_asset` or something like that and not bother with `raw` vs non-raw separation (which is a bit confusing to me at least)?

Reply:
Further bikeshedding of the method name: `update_metadata()`? `modify_metadata()`?

I think the `raw` vs. non-`raw` separation is useful for when you're working with non-model metadata, in which case converting it to a model and back would just be a waste of cycles.  Or are you suggesting one method that takes both raw and non-raw and converts the latter into the former?

Reply:
`.update*` and `.modify*` suggest in-place operations for me, which isn't what happening here, unless we mutate the Asset instance with returned new id etc, which I don't think we should do.

ok, let's keep `raw` if you find it useful.

Reply:
How about `apply_metadata()`?

Reply:
also suggests mutating an existing object.  I think following all the get/set, could be `get_updated`

Reply:
That suggests to me that it's only mutating the local object, not the data on the server.  I really think `set_metadata()` is the best choice.

Reply:
but new metadata produces a new asset, so it would need to mutate the `RemoteAsset` instance then to replace everything (id, metadata) with new metadata as on the remote server.

Reply:
@satra We need another opinion on what to name this method.

Reply:
@yarikoptic I've decided to make the method update the `RemoteAsset` in place and stick with the `set_{raw,}_metadata()` name.

Reply:
Ok, I see advantages to that from UX perspective.  Thank you. Let's see where it leads us ;-)

Reply:
For a further viewpoint: libraries such as [ImmutableJS](https://github.com/immutable-js/immutable-js) have data structure APIs with methods named `update()`, `set_*()`, et al. that actually manufacture new instances and return them, leaving the old instance unchanged. The use of mutational verbs for these actions isn't a bad thing if it's understood that you're working with immutable data. In fact, it's a nice abstraction to be able to say "set the value to X" and hide the fact that nothing is really being set at all because it continues the semantics that we are accustomed to in earlier, non-immutable data structure libraries.

Following an argument like that, I think using mutational names here is perfectly fine. (This argument also applies to using `PUT` to apparently "modify" these resources in the REST API.)

Reply:
Thank you @waxlamp for more input/examples.  It still seats not entirely well with me, but I guess let's proceed the "mutable" way and see where it brings us ;-)